x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptor.java

```diff
@@ -140,6 +194,108 @@ public <T extends TransportResponse> void sendRequest(
+                                + connection.getVersion()
+                                + "] does not support receiving them"
+                        )
+                    );
+                    return;
+                }
+
+                final Authentication authentication = securityContext.getAuthentication();
+                assert authentication != null : "authentication must be present in security context";
+
+                final Optional<String> remoteClusterAlias = remoteClusterAliasResolver.apply(connection);
+                assert remoteClusterAlias.isPresent() : "remote cluster alias must be set for the transport connection";
+
+                final ThreadContext threadContext = securityContext.getThreadContext();
+                final var contextRestoreHandler = new ContextRestoreResponseHandler<>(threadContext.newRestorableContext(true), handler);
+                authzService.retrieveRemoteAccessRoleDescriptorsIntersection(
+                    remoteClusterAlias.get(),
+                    authentication.getEffectiveSubject(),
+                    ActionListener.wrap(roleDescriptorsIntersection -> {
+                        try (ThreadContext.StoredContext ignore = threadContext.stashContext()) {
+                            writeCredentialForClusterToContext(remoteClusterAlias.get(), threadContext);
+                            new RemoteAccessAuthentication(authentication, roleDescriptorsIntersection).writeToContext(threadContext);
+                            sender.sendRequest(connection, action, request, options, contextRestoreHandler);
+                        }
+                    }, e -> contextRestoreHandler.handleException(new SendRequestTransportException(connection.getNode(), action, e)))
+                );
+            }
+
+            private boolean isWhitelistedForRemoteAccessHeaders(final TransportRequest request) {
+                return request instanceof ShardSearchRequest
+                    || request instanceof ShardFetchRequest
+                    || request instanceof SearchRequest
+                    || request instanceof ClusterSearchShardsRequest;
+            }
+
+            private void writeCredentialForClusterToContext(final String remoteClusterAlias, final ThreadContext threadContext) {
+                final String clusterCredential = remoteClusterAuthorizationResolver.resolveAuthorization(remoteClusterAlias);
+                // This can happen if the cluster credential was updated after we made the initial check to send remote access headers
+                // In this case, fail the request. In the future we may want to retry instead, to pick up the updated remote cluster
+                // configuration
+                if (clusterCredential == null) {
+                    throw new IllegalStateException(
+                        "remote cluster credential unavailable for target cluster [" + remoteClusterAlias + "]"
+                    );
+                }
```

Comment:
I wonder whether we should replace `boolean shouldSendWithRemoteAccessHeaders(...)` with something like
`Optional<RemoteAccessHeaders> getRemoteAccessHeaders(...)`. It avoids this problem and remove other duplicate reads as well. I personally also mostly prefer more elaborative code instead of a single large conditional statement, i.e. multiple `if` checks similar to [this](https://github.com/elastic/elasticsearch/blob/9d605b576f47c39ce7d7501a29181849eda6b713/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java#L393). It also provides more opportunity for logging.

Reply:
`Optional<RemoteAccessHeaders> getRemoteAccessHeaders(...)` works and potentially simplifies things. I went with a boolean to keep the `shouldSendWithRemoteAccessHeaders` as single-purpose as possible but I don't feel strongly here.

Even if we go with `Optional<RemoteAccessHeaders> getRemoteAccessHeaders(...)` though, I wonder if we should still re-resolve the credential just before sending -- there is some time between the initial check and getting to the point of issuing the request (i.e., for resolving the roles etc). Resolving the credential again allows us to pick up last minute changes to the setting that would otherwise go undetected. We still don't get a proper lock on the setting though, so I can also see the argument for the simplified flow, where we only fetch it once. WDYT?

Reply:
I prefer fetching it once. I feel a proper solution is somewhere else, e.g. let `Connection` carries it alongside `clusterAlias`. So let's keep it simple here.